lib/acts_as_ar_query.rb

```diff
@@ -0,0 +1,205 @@
+
+  # @param val [Hash]
+  # TODO [Array, Hash, String]
+  def where(*val)
+    val = val.flatten
+    val = val.first if val.size == 1 && val[0].kind_of?(Hash)
+    dup.tap do |r|
+      old_where = r.options[:where]
+      if val.empty?
+        # nop
+      elsif old_where.blank?
+        r.options[:where] = val
+      elsif old_where.kind_of?(Hash) && val.kind_of?(Hash)
+        val.each_pair do |key, value|
+          old_where[key] = if old_where[key]
+                             Array.wrap(old_where[key]) + Array.wrap(value)
+                           else
+                             value
+                           end
+        end
+      else
+        raise ArgumentError,
+              "Need to support #{__callee__}(#{val.class.name}) with existing #{old_where.class.name}"
+      end
+    end
+  end
+
+  # @param val [Array] for includes
+  # TODO: support hash
+  alias includes append_hash_arg
+
+  # @param val Array for includes
+  alias references append_hash_arg
+
+  # @param [Integer] val
+  alias limit assign_arg
+
+  # @param [String, Symbol, Arel] val
+  # TODO: support Hash
+  # TODO: support mixing?
+  alias order append_hash_arg
+  alias group append_hash_arg
+
+  def reorder(*val)
+    val = val.flatten
+    if val.first.kind_of?(Hash)
+      raise ArgumentError, "Need to support #{__callee__}(#{val.class.name})"
+    end
+
+    dup.tap do |r|
+      r.options[:order] = val
+    end
+  end
+
+  def except(*val)
+    val = val.flatten
+    dup.tap do |r|
+      val.each do |key|
+        r.options.delete(key)
+      end
+    end
+  end
+
+  # similar to except. difference being this persists across merges
+  def unscope(*val)
+    val = val.flatten
+    dup.tap do |r|
+      val.each do |key|
+        r.options[key] = nil
+      end
+    end
+  end
+
+  # @param [Integer] val
+  alias offset assign_arg
+
+  # complete
+  # @param val [Array<Sting,Symbol>,String, Symbol]
+  alias select append_hash_arg
+
+  # execution methods
+  # these methods execue the query
+
+  def to_a
+    @results ||= klass.find(:all, legacy_options)
+  end
+
+  def all
+    self
+  end
+
+  def find(mode, options = {})
+    klass.find(mode, legacy_options.merge(options))
+  end
+
+  def count
+    to_a.size
+  end
+  def_delegators :to_a, :size, :take
+
+  def first # (number)
+    defined?(@results) ? @results.first : klass.find(:first, legacy_options)
+  end
+
+  def last
+    defined?(@results) ? @results.first : klass.find(:last, legacy_options)
+  end
+
+  # benind the scenes
+
+  def dup
+    self.class.new(klass, options.dup)
+  end
+
+  def legacy_options
+    {
+      :conditions => options[:where],
+      :include    => options[:includes],
+      # :include  => options[:references],
+      :limit      => options[:limit],
+      :order      => options[:order],
+      :offset     => options[:offset],
+      :select     => options[:select],
+      :group      => options[:group],
+    }.delete_blanks
+  end
+
+  def instances_are_derived?
```

Comment:
we are calling `instances_are_derived?` in the code, so this mimics the behavior - yes, I want to remove.

once this is removed from rbac then we'll see about removing from here